### PR TITLE
Automatically initialize the configured helm home directory

### LINKED ISSUE #139: the 'home' parameter on the helm provider is broken
### Terraform Version
Terraform v0.11.10
+ provider.aws v1.41.0
+ provider.external v1.0.0
+ provider.helm v0.1.0
+ provider.kubernetes v1.3.0
+ provider.local v1.1.0
+ provider.null v1.0.0
+ provider.template v1.0.0

### Affected Resource(s)
- helm provider
- helm repository

### Terraform Configuration Files
works:

```hcl
provider "helm" {
  service_account = "tiller"
  tiller_image    = "gcr.io/kubernetes-helm/tiller:v2.11.0"
  namespace       = "kube-system"
  debug           = "true"

  kubernetes {
    config_path = "./kubeconfig_${var.release_env[var.environment]}"
  }
}

resource "helm_repository" "artifactory" {
  name = "artifactory"
  url  = "https://artifactory.granicuslabs.com/artifactory/helm"
}
```
doesn't work:
```hcl
provider "helm" {
  service_account = "tiller"
  tiller_image    = "gcr.io/kubernetes-helm/tiller:v2.11.0"
  namespace       = "kube-system"
  debug           = "true"
  home            = "./.helm"

  kubernetes {
    config_path = "./kubeconfig_${var.release_env[var.environment]}"
  }
}

resource "helm_repository" "artifactory" {
  name = "artifactory"
  url  = "https://artifactory.granicuslabs.com/artifactory/helm"
}
```

### Expected Behavior
it should initialize helm in the directory specified in 'home' and install tiller.

### Actual Behavior
```
helm_repository.artifactory: Creating...
  metadata.#: "" => "<computed>"
  name:       "" => "artifactory"
  url:        "" => "https://artifactory.granicuslabs.com/artifactory/helm"

Error: Error applying plan:

1 error(s) occurred:

* helm_repository.artifactory: 1 error(s) occurred:

* helm_repository.artifactory: Couldn't load repositories file (.helm/repository/repositories.yaml).
You might need to run `helm init` (or `helm init --client-only` if tiller is already installed)
```

so I run 
```
export HELM_HOME="./.helm"
helm init --client-only    
Creating ./.helm
Creating .helm/repository
Creating .helm/repository/cache
Creating .helm/repository/local
Creating .helm/plugins
Creating .helm/starters
Creating .helm/cache/archive
Creating .helm/repository/repositories.yaml
Adding stable repo with URL: https://kubernetes-charts.storage.googleapis.com
Adding local repo with URL: http://127.0.0.1:8879/charts
$HELM_HOME has been configured at ./.helm.
Not installing Tiller due to 'client-only' flag having been set
Happy Helming!
```

THEN I run `terraform apply` again and get:
```helm_repository.artifactory: Creating...
  metadata.#: "" => "<computed>"
  name:       "" => "artifactory"
  url:        "" => "https://artifactory.granicuslabs.com/artifactory/helm"

Error: Error applying plan:

1 error(s) occurred:

* helm_repository.artifactory: 1 error(s) occurred:

* helm_repository.artifactory: Looks like "https://artifactory.granicuslabs.com/artifactory/helm" is not a valid chart repository or cannot be reached: open .helm/repository/cache/.helm/repository/cache/artifactory-index.yaml: no such file or directory```
```
in addition, helm tiller is not installed in my cluster.

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. add a `home = ./.helm` to the helm provider 
2. `terraform apply`

### References
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:
- GH-1234

### PR DESCRIPTION
**Pull request by pdecat**
This PR aims to automatically initialize the configured helm home directory and resolve #139.

It also eliminates the need to download `helm` and execute `helm init --client-only` during travis CI builds:

```
      script:
        # needed till we fix the init in the provider
        - helm init --client-only
        - travis_wait 60 make testacc
```

cf. https://github.com/terraform-providers/terraform-provider-helm/blob/v0.9.1/.travis.yml#L54

**Notes**:
- This PR depends on the refactoring of the helm init command to avoid shelling out: https://github.com/helm/helm/pull/5189

TODO:

- [x] run acceptance tests
- [x] make all acceptance tests pass
- [x] remove `replace k8s.io/helm => github.com/pdecat/helm v2.11.1-0.20190119161108-a11b965cfd4d+incompatible` from `go.mod` once https://github.com/helm/helm/pull/5189 is merged.

FWIW, I'm currently testing my three recent PRs cumulatively from a [custom branch](https://github.com/pdecat/terraform-provider-helm/tree/test-cumulative) with a real use case.

### CODE REVIEW

**Comment by pdecat** [comment]
This should also resolve https://github.com/terraform-providers/terraform-provider-helm/issues/78

**Comment by pdecat** [comment]
Needs rebase since https://github.com/terraform-providers/terraform-provider-helm/pull/200

**Comment by alexsomesan** [comment]
@pdecat FYI https://github.com/terraform-providers/terraform-provider-helm/pull/200

**Comment by pdecat** [comment]
Yep, noticed 😅

**Comment by pdecat** [comment]
Rebased after #201.

> @hashibot hashibot bot added `size/S` `documentation` and removed `size/XXL` labels a minute ago

:heart: go mod!

**Comment by pdecat** [comment]
Needs rebase since https://github.com/terraform-providers/terraform-provider-helm/pull/221

**Comment by pdecat** [comment]
Helm PR https://github.com/helm/helm/pull/5189 was merged and is targeted for 2.14.0.

**Comment by johngluckmdsol** [comment]
My team would could really use this right now as we are building a containerized API to deploy helm charts from TF.  Is there any ETA as to when this will be merged?

**Comment by pdecat** [comment]
This PR is waiting on helm 2.14 to be released, which will include https://github.com/helm/helm/pull/5189.

**Comment by dminca** [comment]
So, this will do something like

```sh
helm init --service-account tiller
```

by default? Would be awesome to specify the `service-account` name as well :)
All and all, thank you for this one 👍

**Comment by pdecat** [comment]
Acceptance tests pass with minikube 1.0.0 / kubernetes 1.14.1:
```
# make testacc TESTARGS='-count=1'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test "./helm" -v -count=1 -timeout 120m -parallel=4
=== RUN   TestAccDataRepository_basic
--- PASS: TestAccDataRepository_basic (20.23s)
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestAccResourceRelease_basic
=== PAUSE TestAccResourceRelease_basic
=== RUN   TestAccResourceRelease_concurrent
=== PAUSE TestAccResourceRelease_concurrent
=== RUN   TestAccResourceRelease_update
=== PAUSE TestAccResourceRelease_update
=== RUN   TestAccResourceRelease_emptyValuesList
=== PAUSE TestAccResourceRelease_emptyValuesList
=== RUN   TestAccResourceRelease_setStringValues
--- PASS: TestAccResourceRelease_setStringValues (42.73s)
=== RUN   TestAccResourceRelease_updateValues
=== PAUSE TestAccResourceRelease_updateValues
=== RUN   TestAccResourceRelease_updateMultipleValues
=== PAUSE TestAccResourceRelease_updateMultipleValues
=== RUN   TestAccResourceRelease_repository
=== PAUSE TestAccResourceRelease_repository
=== RUN   TestAccResourceRelease_repository_url
=== PAUSE TestAccResourceRelease_repository_url
=== RUN   TestAccResourceRelease_updateAfterFail
=== PAUSE TestAccResourceRelease_updateAfterFail
=== RUN   TestAccResourceRelease_updateExistingFailed
=== PAUSE TestAccResourceRelease_updateExistingFailed
=== RUN   TestAccResourceRelease_updateVersionFromRelease
=== PAUSE TestAccResourceRelease_updateVersionFromRelease
=== RUN   TestGetValues
--- PASS: TestGetValues (0.00s)
=== RUN   TestAccResourceRepository_basic
--- PASS: TestAccResourceRepository_basic (53.86s)
=== CONT  TestAccResourceRelease_basic
=== CONT  TestAccResourceRelease_repository
=== CONT  TestAccResourceRelease_updateExistingFailed
=== CONT  TestAccResourceRelease_emptyValuesList
--- PASS: TestAccResourceRelease_emptyValuesList (39.98s)
=== CONT  TestAccResourceRelease_updateAfterFail
--- PASS: TestAccResourceRelease_basic (42.06s)
=== CONT  TestAccResourceRelease_repository_url
E0424 19:05:21.360196   24852 portforward.go:372] error copying from remote stream to local connection: readfrom tcp4 127.0.0.1:44661->127.0.0.1:48966: write tcp4 127.0.0.1:44661->127.0
.0.1:48966: write: broken pipe
--- PASS: TestAccResourceRelease_updateExistingFailed (81.49s)
=== CONT  TestAccResourceRelease_update
--- PASS: TestAccResourceRelease_updateAfterFail (44.72s)
=== CONT  TestAccResourceRelease_updateMultipleValues
--- PASS: TestAccResourceRelease_repository (86.94s)
=== CONT  TestAccResourceRelease_updateValues
--- PASS: TestAccResourceRelease_update (47.92s)
=== CONT  TestAccResourceRelease_concurrent
--- PASS: TestAccResourceRelease_updateValues (51.40s)
=== CONT  TestAccResourceRelease_updateVersionFromRelease
--- PASS: TestAccResourceRelease_updateMultipleValues (53.93s)
--- PASS: TestAccResourceRelease_repository_url (105.78s)
--- PASS: TestAccResourceRelease_concurrent (44.89s)
--- PASS: TestAccResourceRelease_updateVersionFromRelease (39.32s)
PASS
ok      github.com/terraform-providers/terraform-provider-helm/helm     294.580s
```

**Comment by pdecat** [comment]
Hi @alexsomesan, I believe this PR is ready for review.

I believe the fact that acceptance tests are passing in CI with the [removal of `helm init --client-only`](https://github.com/terraform-providers/terraform-provider-helm/pull/185/files#diff-354f30a63fb0907d4ad57269548329e3L54) is enough.

An additional test case to cover would be when `init_helm_home` is set to `false`. Not decided yet how to implement this one. Maybe defining `helm_home` to a random empty directory and expecting a failure.

**Comment by rporres** [comment]
The change looks great, @pdecat 

If you want to give a try to that test, it would be awesome although I agree that tests passing with the removal of `--client-only` are a very good indicator.

**Comment by pdecat** [comment]
I've added tests for the `init_helm_home` parameter and made all the existing acceptance tests use an empty helm home and not mess with the user's `~/.helm` directory.

**Comment by allantargino** [comment]
Thanks for this PR! I'm willing to use it :)

**Comment by pdecat** [comment]
~I'm going to update this PR to use the [freshly released helm 2.14.0 tag](https://github.com/helm/helm/releases/tag/v2.14.0).~

I'm going to submit another PR just like https://github.com/terraform-providers/terraform-provider-helm/pull/261 and update tiller at the same time.

**Comment by pdecat** [comment]
Will rebase this PR once https://github.com/terraform-providers/terraform-provider-helm/pull/277 is merged.

**Comment by gbergere** [comment]
Hey @pdecat, I'm really looking forward to this.

Any news on it ? 

Many Thanks

**Comment by pdecat** [comment]
Hi @gbergere, this PR is ready IMO.

Just waiting for maintainers to complete the review (ping @alexsomesan @rporres :wink: ).

File: website/docs/index.html.markdown
Comment (alexsomesan) [inline website/docs/index.html.markdown]: Should we mention here that the value of the `home` attribute is used as the newly initialized location?

**Review by alexsomesan — APPROVED** [review]
Looks good! 
Thank you for tackling this issue. This improvement is needed by many users with more complicated / restricted setups.
One small comment on docs, but nothing to block on.

File: website/docs/index.html.markdown
Comment (pdecat) [reply website/docs/index.html.markdown]: Done!

**Comment by alexsomesan** [comment]
I'm happy with the changes.
I'm holding off merging for a bit so @rporres and @meyskens can have a say if anything stands out to them.

**Pull request merged**
